torchio/data/queue.py

```diff
@@ -146,15 +147,39 @@ def __init__(
         self.shuffle_subjects = shuffle_subjects
         self.shuffle_patches = shuffle_patches
         self.samples_per_volume = samples_per_volume
+        if type(samples_per_volume) == int:
+            self.samples_per_volume = [samples_per_volume for _ in range(
+                self.num_subjects)]
+        assert type(self.samples_per_volume) == list, ('`samples_per_volume`'
+                                                       'should be an int'
+                                                       'or a list')
+        assert len(self.samples_per_volume) == self.num_subjects, (
+            '`samples_per_volume` (list) length must be equal'
+            'to the number of subjects')
         self.sampler = sampler
         self.num_workers = num_workers
         self.verbose = verbose
         self._subjects_iterable = None
-        if start_background:
-            self._initialize_subjects_iterable()
         self.patches_list: List[Subject] = []
         self.num_sampled_patches = 0
 
+        # Same random shuffling applied to subjects and volumes
+        if self.shuffle_subjects:
+            tmp_seed = np.random.random()
+            random.Random(tmp_seed).shuffle(self.subjects_dataset._subjects)
```

Comment:
Maybe it's not a good idea to modify the dataset instance. Do you know if there's a way to have a local, shuffled version of the dataset? Maybe using PyTorch `RandomSampler`?

Reply:
Like passing a cloned (shuffled) `self.subjects_dataset` to the DataLoader in `_get_subjects_iterable()` while not touching self.subjects_dataset?

Reply:
I've just sent a new pull request tackling the other comments and trying to do this.